### PR TITLE
Support signals that do not immediately generate a workflow task

### PR DESCRIPTION
**Pull request by pdoerner**
**What changed?**
Added support for the new `SkipGenerateWorkflowTask` flag on signals.

**Why?**
Resolves https://github.com/temporalio/temporal/issues/3406

**How did you test it?**
New unit/integration tests

**Potential risks**

**Is hotfix candidate?**
No

### CODE REVIEW

File: service/history/historyEngine_test.go [hunk 1]
```diff
@@ -1666,6 +1666,82 @@ func (s *engineSuite) TestRespondWorkflowTaskCompletedSingleActivityScheduledWor
...
+			Identity:              identity,
+			ReturnNewWorkflowTask: true,
+		},
+	})
+	s.NoError(err)
+	s.NotNil(resp)
+
+	s.Equal(enumsspb.WORKFLOW_EXECUTION_STATE_RUNNING, ms.GetExecutionState().State)
+
+	return resp
```
Comment (yycptt) [inline service/history/historyEngine_test.go:1742]: I don't quite follow. 
- When handling the signal request, the logic will load mutable state and found the signal is already there I think (dedup based on requestID)?
- I don't see an assert for if a workflow task is generated?

File: service/history/api/signalworkflow/api.go [hunk 2]
```diff
@@ -106,7 +106,8 @@ func Invoke(
 				request.GetSignalName(),
 				request.GetInput(),
 				request.GetIdentity(),
-				request.GetHeader()); err != nil {
+				request.GetHeader(),
+				request.GetSkipGenerateWorkflowTask()); err != nil {
```
Comment (yycptt) [inline service/history/api/signalworkflow/api.go:110]: also need to update the `createWorkflowTask` I think. Can you plz double check the logic for signal with start workflow? I think some update is also needed there.

Please also add some tests for those two apis re. skipping workflow task generation.

**Review by yycptt — COMMENTED** [review]
@yux0 Could you help review the ndc event reapply part?

File: service/history/api/signalworkflow/api.go (hunk 2 above)
Comment (pdoerner) [reply service/history/api/signalworkflow/api.go:110]: On the signal with start, I did update `signal_with_start_workflow#signalWorkflow()` to pass through the new flag. Is there some other change that is also required?

Could you elaborate more on what change is needed for `createWorkflowTask`? I'm not sure I understand what change is needed there.

File: service/history/workflowTaskHandlerCallbacks.go [hunk 3]
```diff
@@ -452,7 +452,8 @@ func (handler *workflowTaskHandlerCallbacksImpl) handleWorkflowTaskCompleted(
 	)
 	// hasPendingUpdates indicates if there are more pending updates (excluding those which are accepted/rejected by this workflow task).
 	hasPendingUpdates := weContext.UpdateRegistry().HasPending(request.GetMessages())
-	hasBufferedEvents := ms.HasBufferedEvents()
+	// hasBufferedEvents indicates if there are any buffered events which should generate a new workflow task
+	hasBufferedEvents := ms.HasBufferedEvents() && ms.HasAnyBufferedEvent(eventShouldGenerateNewTaskFilter)
```
[the remark this answers is not in the document]
Comment (pdoerner) [reply service/history/workflowTaskHandlerCallbacks.go:456]: I think that it is fine if the value is overwritten on L528. `workflowTaskHandler` only updates the value of `hasBufferedEvents` in the case that "timer was cancelled and its TimerFired event was deleted from buffered events" which is behavior I think we want to keep.

I tried only updating L567 first, but it breaks the failed workflow task handling. The `if wtFailedCause != nil` block overwrites `hasBufferedEvents` to `true` on L549 and if this was `false` previously, then `HasAnyBufferedEvent` will always be false and a new task will never get generated.

File: service/history/historyEngine_test.go (hunk 1 above)
Comment (yycptt) [reply service/history/historyEngine_test.go:1742]: I think the point here is that the mutable state you created already contains the signal and SignalWorkflowExecution is again called for the same signal which makes it a no-op. 

So if you just want to the test the buffered event filtering logic, then I think there's no need for the separate signalWorkflowExecution call.

File: service/history/workflow/mutable_state.go [hunk 4]
```diff
@@ -63,6 +63,8 @@ func (policy TransactionPolicy) Ptr() *TransactionPolicy {
 	return &policy
 }
 
+type BufferedEventFilter func(*historypb.HistoryEvent) bool
```
Comment (yux0) [inline service/history/workflow/mutable_state.go:66]: Should it be defined in history builder?

File: service/history/workflow/mutable_state.go [hunk 5]
```diff
@@ -201,6 +203,7 @@ type (
 		IsTransientWorkflowTask() bool
 		ClearTransientWorkflowTask() error
 		HasBufferedEvents() bool
+		HasAnyBufferedEvent(filter BufferedEventFilter) bool
```
Comment (yux0) [inline service/history/workflow/mutable_state.go:206]: I didn't see the caller of this method on my phone screen. Could you point me to the caller?

**Review by yux0 — COMMENTED** [review]
LGTM overall. Could you check this works for replication and add a functional test to cover it in replication(xdc)?

File: service/history/api/create_workflow_util.go [hunk 6]
```diff
@@ -95,6 +95,7 @@ func NewWorkflowWithSignal(
 			signalWithStartRequest.GetSignalInput(),
 			signalWithStartRequest.GetIdentity(),
 			signalWithStartRequest.GetHeader(),
+			signalWithStartRequest.GetSkipGenerateWorkflowTask(),
```
Comment (yycptt) [inline service/history/api/create_workflow_util.go:98]: first workflow task should not be generated if signalWithStartRequest has skip workflow task flag set.

File: service/history/ndc/events_reapplier.go [hunk 7]
```diff
@@ -96,13 +96,16 @@ func (r *EventsReapplierImpl) ReapplyEvents(
 		return nil, serviceerror.NewInternal("unable to reapply events to closed workflow.")
 	}
 
+	shouldScheduleSignalTask := false
```
Comment (yycptt) [inline service/history/ndc/events_reapplier.go:99]: nit: `shouldScheduleSignalTask` -> `shouldScheduleWorkflowTask` ?

File: service/history/workflow/mutable_state.go (hunk 5 above)
Comment (yycptt) [reply service/history/workflow/mutable_state.go:206]: it's used in workflowTaskHandlerCallback.go to see if any bufferedEvent needs a workflow task

File: service/history/api/create_workflow_util.go (hunk 6 above)
Comment (yycptt) [reply service/history/api/create_workflow_util.go:98]: synced offline. workflow task for new workflow should still be generated.

**Review by yycptt — APPROVED** [review]

File: service/history/ndc/events_reapplier.go [hunk 8]
```diff
@@ -96,13 +96,16 @@ func (r *EventsReapplierImpl) ReapplyEvents(
 		return nil, serviceerror.NewInternal("unable to reapply events to closed workflow.")
 	}
 
+	shouldScheduleWorkflowTask := false
```
Comment (yycptt) [inline service/history/ndc/events_reapplier.go:99]: @yux0 I found some [duplicated logic](https://github.com/temporalio/temporal/blob/master/service/history/api/reapplyevents/api.go#L182) for scheduling a new workflow task in the api handler for ReapplyEvents. To me it seems like this eventsReapplier already takes care of scheduling new workflow task so in the api handler, we should always use false for the `CreateWorkflowTask` field of update workflow post action.

Comment (yycptt) [reply service/history/ndc/events_reapplier.go:99]: Synced with @yux0 offline, we should removed the above-linked duplicated logic for scheduling new workflow task in ReapplyEvents api handler.
@pdoerner you can decide if you want to do this in a separate PR.

**Comment by yux0** [comment]
The reapply part looks good to me.

**Review by yux0 — APPROVED** [review]

File: service/history/api/reapplyevents/api.go [hunk 9]
```diff
@@ -195,13 +187,10 @@ func Invoke(
...
-				return &api.UpdateWorkflowAction{
-					Noop:               true,
-					CreateWorkflowTask: false,
-				}, nil
-			}
-			return postActions, nil
+			return &api.UpdateWorkflowAction{
+				Noop:               true,
+				CreateWorkflowTask: false,
+			}, nil
```
Comment (wxing1292) [inline service/history/api/reapplyevents/api.go:193]: this section of logic will skip the persistence of mutable state if events are reapplied?

**Pull request merged**